src/UhdmAst.cpp

```diff
@@ -1973,76 +1985,94 @@ AstNode* visit_object(vpiHandle obj_h, UhdmShared& shared) {
         return process_param_assign(obj_h, shared);
     }
     case vpiInterface: {
-        // Interface definition is represented by a module node
-        AstIface* elaboratedInterface = new AstIface(make_fileline(obj_h), objectName);
-        bool hasModports = false;
-        visit_one_to_many(
-            {
-                vpiPort,
-                vpiParamAssign,
-                vpiInterfaceTfDecl,
-                vpiModPath,
-                vpiContAssign,
-                vpiInterface,
-                vpiInterfaceArray,
-                vpiProcess,
-                vpiGenScopeArray,
+        std::string modType = get_object_name(obj_h, {vpiDefName});
+        AstIface* interfacep = nullptr;
 
-                // from vpiInstance
-                vpiProgram,
-                vpiProgramArray,
-                vpiTaskFunc,
-                vpiArrayNet,
-                vpiSpecParam,
-                vpiAssertion,
-                vpiNet,
-            },
-            obj_h, shared, [&](AstNode* port) {
-                if (port) { elaboratedInterface->addStmtp(port); }
+        // Check if we have encountered this object before
+        auto it = shared.top_nodes.find(modType);
+        if (it != shared.top_nodes.end()) {
+            // Was created before, fill missing
+            interfacep = reinterpret_cast<AstIface*>(it->second);
+            bool hasModports = false;
+            visit_one_to_many(
+                {
+                    vpiPort,
+                    vpiParamAssign,
+                    vpiInterfaceTfDecl,
+                    vpiModPath,
+                    vpiContAssign,
+                    vpiInterface,
+                    vpiInterfaceArray,
+                    vpiProcess,
+                    vpiGenScopeArray,
+
+                    // from vpiInstance
+                    vpiProgram,
+                    vpiProgramArray,
+                    vpiTaskFunc,
+                    vpiArrayNet,
+                    vpiSpecParam,
+                    vpiAssertion,
+                    vpiNet,
+                    vpiVariables
+                },
+                obj_h, shared, [&](AstNode* portp) {
+                    if (portp) { interfacep->addStmtp(portp); }
+                });
+            visit_one_to_many({vpiModport}, obj_h, shared, [&](AstNode* portp) {
+                if (portp) {
+                    hasModports = true;
+                    interfacep->addStmtp(portp);
+                }
             });
-        visit_one_to_many({vpiModport}, obj_h, shared, [&](AstNode* port) {
-            if (port) {
-                hasModports = true;
-                elaboratedInterface->addStmtp(port);
+            if (hasModports) {
+                // Only then create the nets, as they won't be connected otherwise
+                visit_one_to_many({vpiNet}, obj_h, shared, [&](AstNode* portp) {
+                    if (portp) { interfacep->addStmtp(portp); }
```

Comment:
This has the same problem as the original code - nets are visited in line 2016 (and added to interface), then potentially added here again.
How does it behave with i.e. [OneNetModport test](https://github.com/chipsalliance/UHDM-integration-tests/blob/master/tests/OneNetModport/top.sv)? Does it create duplicate nodes? Are both visits needed?

Reply:
I removed the second visit.